src/Utils/Traceroute.ts

```diff
@@ -0,0 +1,236 @@
+import * as raw from 'raw-socket';
+import * as dgram from 'dgram';
+import { promises as dns } from 'dns';
+import { EventEmitter } from 'events';
+import { isIP } from 'net';
+
+export { Protocol } from 'raw-socket';
+
+type Options = {
+  amountProbes: number;
+  maximumHops: number;
+  timeoutInMillis: number;
+  reverseLookup: boolean;
+  protocol: number;
+  packetSize: number;
+  outStream: typeof process.stdout;
+};
+
+const defaultOptions: Options = {
+  amountProbes: 1,
+  maximumHops: 64,
+  timeoutInMillis: 3000,
+  reverseLookup: true,
+  protocol: raw.Protocol.UDP,
+  packetSize: 52,
+  outStream: process.stdout
+};
+
+export class Traceroute extends EventEmitter {
+  private readonly icmpSocket = raw.createSocket({
+    protocol: raw.Protocol.ICMP
+  });
+  private readonly resolver = new dns.Resolver();
+  private readonly options: Options;
+  private port = 33433;
+  private ttl = 1;
+  private startTime?: [number, number];
+  private probes = 0;
+  private timeout?: NodeJS.Timeout;
+  private previousIP?: string;
+  private udpSocket?: dgram.Socket;
+  private destinationHostname: string;
+
+  constructor(
+    private destinationIp: string,
+    userOptions: Partial<Options> = {}
+  ) {
+    super();
+    this.options = { ...defaultOptions, ...userOptions };
+    this.destinationHostname = this.destinationIp;
+
+    this.icmpSocket.on('message', async (buffer: Buffer, ip: string) => {
+      const port = this.udpSocket
+        ? buffer.readUInt16BE(50)
+        : buffer.readUInt16BE(buffer.length - 2);
+      if (port === this.port) {
+        const hostName = await this.getHostName(ip);
+        this.handleReply(ip, hostName);
+      }
+    });
+  }
+
+  public async start(): Promise<void> {
+    if (!isIP(this.destinationIp)) {
+      this.destinationIp = (
+        await this.resolver.resolve(this.destinationHostname, 'A')
+      )[0];
+    }
+
+    process.stdout.write(
+      `traceroute to ${this.destinationHostname} (${this.destinationIp}), ${this.options.maximumHops} hops max, ${this.options.packetSize} byte packets`
+    );
+
+    if (this.options.protocol === raw.Protocol.UDP) {
+      this.udpSocket = dgram.createSocket('udp4');
+      this.udpSocket.bind(() => this.sendPacket());
```

Comment:
If binding fails, an 'error' event is generated. Please add a listener for it (see comments above)